lib/src/git.rs

```diff
@@ -60,6 +60,38 @@ fn local_branch_name_to_ref_name(branch: &str) -> String {
     format!("refs/heads/{branch}")
 }
 
+fn resolve_git_ref_to_commit_id(
+    git_ref: &git2::Reference<'_>,
+    known_target: Option<&RefTarget>,
+) -> Option<CommitId> {
+    // Try fast path if we have a candidate id which is known to be a commit object.
+    if let Some(RefTarget::Normal(id)) = known_target {
+        if matches!(git_ref.target(), Some(oid) if oid.as_bytes() == id.as_bytes()) {
+            return Some(id.clone());
+        }
+        if matches!(git_ref.target_peel(), Some(oid) if oid.as_bytes() == id.as_bytes()) {
+            // A hard tag. No need to peel to a tag object.
+            return Some(id.clone());
+        }
+        // A tag (according to ref name.) Try to peel one more level. This is slightly
+        // faster than recurse into peel_to_commit(). If we recorded a tag oid, we
+        // could skip this at all.
+        if let Some(Ok(tag)) = git_ref.is_tag().then(|| git_ref.peel_to_tag()) {
+            if tag.target_id().as_bytes() == id.as_bytes() {
+                // No need to peel recursively to a commit object.
+                return Some(id.clone());
+            } else {
+                // Unknown id. Continue peeling from the tag object.
+                let git_commit = tag.into_object().peel_to_commit().ok()?;
```

Comment:
(Very optional): Coming from a person who doesn't really understand what `peel_to_commit` does, I find this part of the code confusing. Perhaps this part of the optimization is not worth the complexity? I'd also appreciate a comment about how this can come about and what "Continue peeling from the tag object" means.

It's entirely up to you, though, as it's quite possible that my questions all have to do with me not understanding the prerequisite background.

Reply:
SG. I also think the new comment is quite helpful; thanks for adding it!